### ISSUE TITLE
server side debugger, fuzzy comprehension of params delivery to controller, fuzzy documentation on returns from ActiveRecord methods

### ISSUE THREAD

**Issue by EricLScace** [ga-wdi-boston/full-stack-project#921]
I know some mention was made of a server-side debug tool, but I cannot find this in my notes anywhere. I've been operating blind this weekend, which has made for some frustrating experiences in trying to understand what's happening when, e.g., a xxx.find_by instruction returns an ambiguous thing (is it a hash? an array of hashes? a list?)

Ditto for parameter delivery from an incoming GET, PATCH, etc: the method documentation is vague — which makes it difficult to test for e.g., absence of an optional field, or form in which an parameter is delivered. Even the proper Ruby syntax for referencing params is being discovered only by guessing — which is way too time-consuming a way to develop.

Basically these 3 issues have cost me way too much time this weekend, and I feel pretty behind schedule now. :-(

**Comment by EricLScace** [comment]
Just come find me when you have a couple of minutes to guide me. I'll post what works here before closing the issue.

**Comment by scottyscripts** [comment]
As far as debugging goes, if you put the line `binding.pry` in your controller, a REPL session will be opened in pry, and code execution will stop so you can take a look at things. You can also `puts` values and will be able to see them in your server window. 

With ActiveRecord, we can start off with a general idea of what will be returned by taking a look at the documentation for methods http://guides.rubyonrails.org/v2.3.11/active_record_querying.html.

As far as params, take a look [here from the Rails guide](http://guides.rubyonrails.org/action_controller_overview.html#parameters), it provides a pretty great explanation of the `params` object.

**Comment by EricLScace** [comment]
Hmm... puts was not working for me. Also difficult to put a variable when you don't know what it is (hash, etc).

**Comment by danman01** [comment]
Using `logger`:
```
logger.debug "Person attributes hash: #{@person.attributes.inspect}"
logger.info "Processing the request..."
logger.fatal "Terminating application, raised unrecoverable error!!!"
```
 and using `binding.pry` are your best friends for debugging. More: http://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-byebug-gem (they reference the `byebug` gem, which is similar to debugging using `pry`)

**Comment by EricLScace** [comment]
Thanks. binding.pry removed the blinders on WTF was happening.

**Comment by danman01** [comment]
@EricLScace can this issue be closed? Thanks!
